100 year plan: New/Existing site flow

aneeshd16 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes Automattic/martech#2184
P2: pau2Xa-5bu-p2

## Proposed Changes

* This PR adds the new/existing site flow for the 100 year plan.

| Step | Desktop | Mobile |
|--------|--------|--------|
| New or Existing Site |<img width="1907" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/73b8a754-5892-4642-8b8a-8979b122de99"> | <img width="311" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/09d3ea83-7fa4-421e-b657-9ae2c441c83b"> |
| Site Picker |<img width="1907" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/6de761c5-6571-47dd-bb44-94d909a096a6"> | <img width="366" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/2787b14d-b377-4049-87f3-b2c861b4db5c"> |
| Site Picker Modal | <img width="1904" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/b3eea66e-4f5f-42ed-aa4e-5a9ad563b9f3"> | <img width="368" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/cb8069ba-5403-4f14-b013-7fd833780b01"> | 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Go to /setup/hundred-year-plan
- If the user has 0 existing sites:
  - The Site title/description step should be shown. The site title field is mandatory.)
  - The next step should be the domains step. The prices for the domains should be hidden. A *.wordpress.com domain suggestions should not be visible.
  - Selecting a domain should lead to checkout.
- If the user has 1 or more existing sites:
  - The New/Existing site step should be shown.
  - Clicking on the “New Site” CTA should start the new-site flow listed above.
  - Clicking on the “Existing Site” CTA should lead to the site picker step.
  - The site picker step should show a list of the user’s sites and should exclude <TBD>.
  - Selecting an existing site should show a confirmation modal.
  - Clicking on the "Chat with us" link should open up the chat window.
  - Clicking on the “Confirm” CTA in the modal should lead to checkout.
- The pre-Checkout loading screen should be according to the design (“Setting up your legacy…”).

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Commit:
100 year plan: New/Existing site flow

Commit:
Fix styling

southp commented:
For a record, I've checked the happy paths covered and they all work as expected:

* An existing user with 0 site
* An existing user with existing sites
    * create a new site
    * choose an existing site
* Also, just in case, the new site flow. 

I haven't had a chance to look into code yet and need to step out, but will focus on identify if there is any edge case and code review later :)

dzver commented:
With a user with only one site, the mobile site selector shows no sites for me

<img width="409" alt="Screenshot 2023-09-15 at 10 24 37" src="https://github.com/Automattic/wp-calypso/assets/82778/c4251faa-acf7-4f5b-8b1c-fdca5b990901">

The desktop flow works fine from start to finish.

southp commented on client/landing/stepper/declarative-flow/hundred-year-plan.ts:
```diff
@@ -68,7 +104,44 @@ const HundredYearPlanFlow: Flow = {
+						product_slug: PLAN_100_YEARS,
+					},
+				];
+				if ( providedDependencies?.selectedDomain ) {
```
@aneeshd16 I'm not sure how I can run through a flow that covers this part of the code since there is no domain step in the existing site flow. Could you help me clarify?

aneeshd16 replied:
You can test this by choosing an existing site with a custom bundled domain. Here's an example:

https://github.com/Automattic/wp-calypso/assets/5436027/c2848b67-958f-4612-b950-a91041379e77

aneeshd16 replied:
@southp, I just realized that adding the site domain to the cart would be incorrect as we don't currently have a flow for renewing existing site domains. I'll remove this code shortly.

southp reviewed (commented):
When a longer domain name is used, the text can overflow horizontally:

<img width="382" alt="截圖 2023-09-15 下午4 44 03" src="https://github.com/Automattic/wp-calypso/assets/1842898/3ca4f191-ed8c-46a8-808a-d13a7d99f0c1">


However, from my view the only critical thing is the disappearing site picker on mobile that @dzver raised above. This can be a follow-up :)

Commit:
Removed BBE check

Commit:
Do not add existing site domain to the cart

Commit:
Adjust max-height for site selector on small screens to prevent overflow and ensure all sites are visible.

Commit:
Capital P dangit

aneeshd16 commented:
Thanks @southp for the review!

> However, from my view the only critical thing is the disappearing site picker on mobile that @dzver raised above. This can be a follow-up :)

I added the follow-up task here: https://github.com/Automattic/martech/issues/2199

> With a user with only one site, the mobile site selector shows no sites for me

I have fixed this via https://github.com/Automattic/wp-calypso/pull/81723/commits/52ed271bd1a4d62221610ec0f5ba1502c3b4677b. The site should be visible now:
<img width="574" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/786c7ad0-de71-4c90-a81a-2a0ca613800f">

dzver reviewed (approved):
WFM now, thanks for the fixes!

southp reviewed (approved):
Thanks for the fix. 

I've encountered two warnings that look important, but let's create follow-up issues for them :) 

On picking an existing site, the Track `calypso_signup_actions_submit_step` event is provided with a non-appropriate prop:
<img width="1032" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1842898/213a3387-df65-4dc3-8ba5-ffa96d0b373d">

On entering the site selector, there is a bad `setState()` call which is worth investigating:
<img width="382" alt="截圖 2023-09-15 下午4 44 03" src="https://github.com/Automattic/wp-calypso/assets/1842898/b969426e-2825-4e2c-af00-56d6679b3175">

Commit:
Submit siteId and siteSlug instead of the site object

aneeshd16 commented:
Thanks @southp for the review!

> On picking an existing site, the Track calypso_signup_actions_submit_step event is provided with a non-appropriate prop:

Nice catch! I have fixed it via https://github.com/Automattic/wp-calypso/pull/81723/commits/e548365f3c61a5f12129230ed2001d60f70f27cb

> On entering the site selector, there is a bad setState() call which is worth investigating:

I noticed that this error is present in all Stepper flows, so we can investigate this separately. Created a follow-up issue here: https://github.com/Automattic/martech/issues/2212

southp reviewed (approved):
The fix LGTM. Thanks for addressing it 🙇🏼

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8985024

Thank you @aneeshd16 for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8985024).
